### PR TITLE
Decoupling Authentication Flow: Mark XXV

### PR DESCRIPTION
**Pull request by jleandroperez**
### Details:
This PR decouples SignupService from the host app:

- `WPAppAnalytics` calls have been replaced by the WordPressAuthenticator relay mechanism.
- We've split up the SignupService sequence, so that the "default account setup" and "sync'ing" is done in the client app.

### Testing:
1. Fresh Install
2. Signup using a Google Account that still has no WordPress.com linked account. Verify everything is pretty!
3. Fresh install again!
4. Attempt to signup using the same Google Account you've used in (2).

@nheagy may i bug you with this one?. It's almost done!!! Wooooo

Thanks in advance!!

### CODE REVIEW

File: WordPress/Classes/Services/SignupService.swift [hunk 1]
```diff
@@ -1,84 +1,74 @@
...
     /// - Parameters:
-    ///   - token: the token from a successful Google login
+    ///   - googleToken: the token from a successful Google login
     ///   - success: block called when account is created successfully
     ///   - failure: block called when account creation fails
-    func createWPComUserWithGoogle(token: String,
-                                   success: @escaping SignupSocialSuccessBlock,
-                                   failure: @escaping SignupFailureBlock) {
+    ///
+    func createWPComUser(googleToken: String,
```
Comment (nheagy) [inline WordPress/Classes/Services/SignupService.swift]: Personally I like `createWPComUserWithGoogle` better 😜 

Either way, renaming this method doesn't seem critical to decoupling. It'd make reviews easier if I didn't have to consider so many changes 😬 also, if we're renaming things changing `SignupService` to mention `Google` would make a lot of sense (as mentioned in #9109)

File: WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift [hunk 2]
```diff
@@ -25,80 +25,129 @@ class SignupGoogleViewController: LoginViewController {
...
-        }
+            self?.socialSignupDidFail(with: error)
+        })
+    }
+
+
+    /// Social Signup Successful: Analytics + Pushing the Signup Epilogue.
+    ///
+    func socialSignupWasSuccessful(with credentials: WordPressCredentials) {
+        WordPressAuthenticator.track(.createdAccount, properties: ["source": "google"])
```
Comment (nheagy) [inline WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift]: This is is a change for the `.createdAccount` event; previously it did not rely on the syncing to be complete. Also, previously it was calling  `BlogSyncFacade().syncBlogs` and this ultimately calls `WordPressComSyncService().service.syncWPCom()` via the delegate.

Also I don't think any sync was being called in the event that the account was new (as there's nothing to sync), so this is new behaviour—and I think it may be incorrect. Can you double check?

File: WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift [hunk 3]
```diff
@@ -25,80 +25,129 @@ class SignupGoogleViewController: LoginViewController {
...
+        WordPressAuthenticator.track(.signupSocialSuccess)
+
+        showSignupEpilogue(for: credentials)
+    }
+
+    /// Social Login Successful: Analytics + Pushing the Login Epilogue.
+    ///
+    func socialLoginWasSuccessful(with credentials: WordPressCredentials) {
+        WordPressAuthenticator.track(.loginSocialSuccess)
```
Comment (nheagy) [inline WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift]: `.loginSocialSuccess` should happen first. I also have similar concerns as with `.createdAccount`, in that previously this didn't rely on the completion of syncing.

File: WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift [hunk 4]
```diff
@@ -25,80 +25,129 @@ class SignupGoogleViewController: LoginViewController {
...
+    func socialSignupWasSuccessful(with credentials: WordPressCredentials) {
+        WordPressAuthenticator.track(.createdAccount, properties: ["source": "google"])
+        WordPressAuthenticator.track(.signupSocialSuccess)
+
+        showSignupEpilogue(for: credentials)
+    }
+
+    /// Social Login Successful: Analytics + Pushing the Login Epilogue.
+    ///
+    func socialLoginWasSuccessful(with credentials: WordPressCredentials) {
```
Comment (nheagy) [inline WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift]: `socialLoginWasSuccessful` is a bit misleading, because we didn't expect to login. Something like `wasLoggedInInstead` explains what happened a bit better (tho it's an attrocious method name) 😜 

Also we don't _know_ if the call was successful. We just know it's complete 😉

File: WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift [hunk 5]
```diff
@@ -25,80 +25,129 @@ class SignupGoogleViewController: LoginViewController {
...
+        WordPressAuthenticator.track(.signupSocialToLogin)
+
+        showLoginEpilogue(for: credentials)
+    }
+
+    /// Social Signup Failure: Analytics + UI Updates
+    ///
+    func socialSignupDidFail(with error: Error) {
+        WPAnalytics.track(.signupSocialFailure)
+
```
Comment (nheagy) [inline WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift]: This is missing the `guard let error = error else { self?.navigationController?.popViewController(animated: true)` bit.

File: WordPress/Classes/Services/SignupService.swift [hunk 6]
```diff
@@ -1,84 +1,74 @@
...
 
+// MARK: - Private
+//
+private extension SignupService {
 
-    // MARK: Private Instance Methods
-
-    @objc func anonymousApi() -> WordPressComRestApi {
-        return WordPressComRestApi(userAgent: WPUserAgent.wordPress())
+    private var anonymousAPI: WordPressComRestApi {
```
Comment (nheagy) [inline WordPress/Classes/Services/SignupService.swift]: is this private redundant, given it's already in a `private extension`?

File: WordPress/Classes/Services/SignupService.swift [hunk 7]
```diff
@@ -1,84 +1,74 @@
...
         static let createdAccount = "created_account"
     }
 }
+
+
+// MARK: - Errors
+//
+extension SignupService {
+
+    enum ServiceError: Error {
```
Comment (nheagy) [inline WordPress/Classes/Services/SignupService.swift]: This exists three times in the code. Any reason not to move it to be defined just once?

File: WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift [hunk 8]
```diff
@@ -7,7 +7,7 @@ class SignupGoogleViewController: LoginViewController {
     // MARK: - Properties
 
     private var hasShownGoogle = false
-    @IBOutlet var titleLabel: UILabel?
+    @IBOutlet var titleLabel: UILabel!
```
Comment (nheagy) [inline WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift]: I bet you could decouple this as-is without changing the `UILabel?` to `UILabel!` 😜

File: WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift [hunk 9]
```diff
@@ -25,80 +25,129 @@ class SignupGoogleViewController: LoginViewController {
...
+        guard !hasShownGoogle else {
+            return
         }
+
+        displayGoogleSingleSignOn()
+        hasShownGoogle = true
     }
 
-    private func showGoogleScreen() {
+    private func displayGoogleSingleSignOn() {
```
Comment (nheagy) [inline WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift]: Let's keep `showGoogleScreen `. I don't think `displayGoogleSingleSignOn` is strictly accurate, anyway. The library is `Google Sign-In`

File: WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift [hunk 10]
```diff
@@ -25,80 +25,129 @@ class SignupGoogleViewController: LoginViewController {
...
         // Flag this as a social sign in.
-        loginFields.meta.socialService = SocialServiceName.google
+        loginFields.meta.socialService = .google
 
         // Configure all the things and sign in.
-        GIDSignIn.sharedInstance().delegate = self
-        GIDSignIn.sharedInstance().uiDelegate = self
-        GIDSignIn.sharedInstance().clientID = WordPressAuthenticator.shared.configuration.googleLoginClientId
-        GIDSignIn.sharedInstance().serverClientID = WordPressAuthenticator.shared.configuration.googleLoginServerClientId
+        guard let googleSSO = GIDSignIn.sharedInstance() else {
```
Comment (nheagy) [inline WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift]: I don't think this bit is necessary for decoupling. is it?

File: WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift [hunk 11]
```diff
@@ -25,80 +25,129 @@ class SignupGoogleViewController: LoginViewController {
...
+        createWordPressComUser(googleUser: googleUser, googleToken: googleToken, googleEmail: googleEmail)
+    }
+}
+
 
+// MARK: - WordPress.com Account Creation Methods
+//
+private extension SignupGoogleViewController {
+
+    /// TODO: Not cool with this. Let's refactor LoginFields, when time permits.
```
Comment (nheagy) [inline WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift]: Let's leave all these changes for a different PR?

File: WordPress/Classes/Services/SignupService.swift [hunk 12]
```diff
@@ -1,84 +1,74 @@
...
 
 
-typealias SignupSocialSuccessBlock = (_ newAccount: Bool) -> Void
-typealias SignupFailureBlock = (_ error: Error?) -> Void
-
-
-/// SignupService is responsible for creating a new WPCom user and blog.
-///
-open class SignupService: LocalCoreDataService {
+// MARK: - SignupService: Responsible for creating a new WPCom user and blog.
```
[the remark this answers is not in the document]
Comment (jleandroperez) [reply WordPress/Classes/Services/SignupService.swift]: This one falls, a bit, in a gray area. Reason why i've used `MARK` (since ever!) is because in a single file you may have more than one class / structure definition.

Specially lately: we use *extensions* quite a lot, to declare protocol conformances, and so on. And using `MARK` is quite useful, because it's *VERY* easy to pinpoint an extension / nested type within Xcode's dropdown:

File: `NotificationRange.swift`

screen shot 2018-04-23 at 2 54 50 pm

--

I think we need the whole team to settle on a documentation standard, because at this precise moment... we've got different styles, and no solid guidelines, i'm afraid!

File: WordPress/Classes/Services/SignupService.swift (hunk 1 above)
Comment (jleandroperez) [reply WordPress/Classes/Services/SignupService.swift]: Swift style would probably be...

```
func createWPComUser(with googleToken:
```

But you're right, this isn't a requirement for decoupling. I was in the neighborhood and felt like improving things a bit.

For the sake of shrinking this PR, i'll revert this change.

[Reference here](https://swift.org/documentation/api-design-guidelines/#strive-for-fluent-usage).

File: WordPress/Classes/Services/SignupService.swift (hunk 6 above)
Comment (jleandroperez) [reply WordPress/Classes/Services/SignupService.swift]: Reasoning behind the double private is:

```
class Test {
    func doSomething() {
        accessible()
        inaccessible() // This should break
    }
}

private extension Test {

    func accessible() {   }

    private func inaccessible() { }
}
```

This was **TRUE** the last time i've checked (Swift 4.0). Now... in the latest compiler, the code posted above doesn't break anymore. So yeah, let's nuke the double private, since ... the compiler changed it's behavior!

File: WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift (hunk 8 above)
Comment (jleandroperez) [reply WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift]: We're soooooo gonna have a long, very, so very long chat on the subject Nate..!!. But you're right, this isn't a decoulping-Y thing, so... for now, we'll do it your way.

**For now** 😃

File: WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift (hunk 9 above)
Comment (jleandroperez) [reply WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift]: Sounds good!

File: WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift (hunk 10 above)
Comment (jleandroperez) [reply WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift]: Not really needed, but we're saving 5 method calls, and replacing them with a single local reference

File: WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift (hunk 3 above)
Comment (jleandroperez) [reply WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift]: Updating, thank you!

File: WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift (hunk 4 above)
Comment (jleandroperez) [reply WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift]: Sounds reasonable, updating!

File: WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift (hunk 11 above)
Comment (jleandroperez) [reply WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift]: Yes sir! that's the idea. Let's, instead, file an issue to do so, which is more ... tracking friendly?.
(And i'll nuke this comment).

Thank you!

File: WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift (hunk 5 above)
Comment (jleandroperez) [reply WordPress/Classes/ViewRelated/NUX/SignupGoogleViewController.swift]: Since there is now an `Unknown` error... we would just get the `Google Sign up failed` legend (which was the idea!).

But still, you're right, i've changed the behavior. I'll add a guard to preserve the original handling, thank you!

File: WordPress/Classes/Services/SignupService.swift (hunk 12 above)
[the remark this answers is not in the document]
Comment (nheagy) [reply WordPress/Classes/Services/SignupService.swift]: I understand what the `MARK` does. If you think it's useful, _adding_ one is fine, but I don't understand why we'd remove the existing documenting comment which can be used elsewhere in the codebase to understand what a type or method does. Looks like this:

screen shot 2018-04-23 at 1 02 38 pm

`MARK` makes browsing the code easier, but the removed comment makes _understanding_ the code easier.

[the remark this answers is not in the document]
Comment (jleandroperez) [reply WordPress/Classes/Services/SignupService.swift]: Ahhh i've **CHANGED** a comment, not added a new one. You're right, reverting right away, thanks for raising this!!

**Comment by jleandroperez** [comment]
@nheagy sir, original behavior restored. Ready for another peek!.

Thanks **A LOT** for your time!!!

**Comment by jleandroperez** [comment]
Thanks A LOOOOOT Nate!!!
